FT2-37-4: Completed the PHP Basic Assignment 4

MD-Aftab-Innoraft opened the pull request:
Description: Created a form which contains: 

1. Two input fields to input 'First name' and 'Last name' respectively.
2. Another text input field namely 'Full name' is disabled by default and displays the sum of 'First name' and 'Last name' when both the fields are filled.
3. A file input field accepts an image.
4. A textarea to input subject and marks in the format 'subject|marks'. Each such pair is on a new line.
5. A text field to input a number and verify it follows the syntax of an Indian 'Phone Number'.
6. A submit button which submits the form. On submission, the form 'Full name', the uploaded image, 'subject|marks' table and  'Phone number' are displayed.

MD-Aftab-Innoraft committed:
Title: Completed the PHP Basic Assignment 4
Description: Two input fields to input 'First name' and 'Last name' respectively.
Another text input field namely 'Full name' is disabled by default and displays the sum of 'First name' and 'Last name' when both the fields are filled
A file input field accepts an image
A textarea to input subject and marks in the format 'subject|marks'. Each such pair is on a new line.
A text field to input a number and verify it follows the syntax of an Indian number
A submit button which submits the form.

MD-Aftab-Innoraft committed:
Added the files again

kumareshinnoraft commented on imageForm.php:
```diff
@@ -0,0 +1,60 @@
+
+<?php
+
+class Form
```
We should add multiline comments for the `class`.

kumareshinnoraft commented on imageForm.php:
```diff
@@ -0,0 +1,60 @@
+
+class Form
+{
+    private $fname;
+    private $lname;
+    private $fullname;
+
```
We should have multiline comments for the `class variables` & `functions`.

kumareshinnoraft commented on index.php:
```diff
@@ -0,0 +1,169 @@
+
+    if ($_SERVER["REQUEST_METHOD"] == "POST") {
+
+        if (empty($_POST["fname"])) {
+            $myForm->fnameErr = "First Name is required";
+        } else {
+            $fname = $_POST["fname"];
+            // check if name only contains letters and whitespace
+            if (!preg_match("/^[a-zA-Z-' ]*$/", $fname)) {
+                $myForm->fnameErr = "Only letters and white space allowed";
+            } else {
+                $myForm->setFirstName($fname);
+            }
+        }
+
+        if (empty($_POST["lname"])) {
+            $myForm->lnameErr = "Last Name is required";
+        } else {
+            $lname = $_POST["lname"];
+            // check if name only contains letters and whitespace
+            if (!preg_match("/^[a-zA-Z-' ]*$/", $lname)) {
+                $myForm->lnameErr = "Only letters and white space allowed";
+            } else {
+                $myForm->setLastName($lname);
+            }
```
Couldn't we move this to a different function?

kumareshinnoraft commented on index.php:
```diff
@@ -0,0 +1,169 @@
+
+    // at this point all the properties are set to default values by constructor
+
+    if ($_SERVER["REQUEST_METHOD"] == "POST") {
+
+        if (empty($_POST["fname"])) {
```
Could we merge this two `if` statement?

MD-Aftab-Innoraft replied:
Sorry, but I didn't get this. On line 10, I have check whether the form has been submitted whereas on line 12 and henceforth(line 24, line 38 and others), I have checked for the values of different input fields when the form is submitted. @kumareshinnoraft please guide me on the same.

kumareshinnoraft commented on index.php:
```diff
@@ -0,0 +1,169 @@
+    <!-- INDEX.PHP  -->
+    <?php
+    require 'imageForm.php';
+
+    // creating a new 'Form' object
+    $myForm = new Form();
+
+    // at this point all the properties are set to default values by constructor
```
We should add comments with capital letter and end it with a full stop.

kumareshinnoraft commented on index.php:
```diff
@@ -0,0 +1,169 @@
+    <!-- INDEX.PHP  -->
+    <?php
```
We should add a space after starting `<?php` code.

kumareshinnoraft commented on index.php:
```diff
@@ -0,0 +1,169 @@
+    <!-- INDEX.PHP  -->
```
We do not require this comment here.

`    <!-- INDEX.PHP  -->
`

kumareshinnoraft commented on index.php:
```diff
@@ -0,0 +1,169 @@
+        if ($myForm->fnameErr == "" && $myForm->lnameErr == "") {
+            $myForm->setFullname();
+        }
+        //    seeing the files in $_FILES superglobal
+        //    echo "<pre>";
+        //    print_r($_FILES);
+        //    echo "</pre>";
```
We should not push commented code in the PR.

kumareshinnoraft commented on index.php:
```diff
@@ -0,0 +1,169 @@
+        $img_name = $_FILES['image']['name'];
+        $img_tmp = $_FILES['image']['tmp_name'];
+        // move $img_tmp to the required folder with the name with which it is saved
+        move_uploaded_file($img_tmp, "uploads/$img_name");
```
We should add an exception handing.

kumareshinnoraft commented on index.php:
```diff
@@ -0,0 +1,169 @@
+        if($subjectmarks == "") {
+            $subjectmarksErr = "No subject|marks pair entered";
+        }
+        else {
+            if(preg_match($subjectmarkspattern, $subjectmarks)) {
```
Couldn't we use `else if` here?

kumareshinnoraft commented on index.php:
```diff
@@ -0,0 +1,169 @@
+
+        $subjectmarks = $_POST["sub_marks"];
+        $subjectmarksErr = "";
+        $subjectmarkspattern = '/^[a-zA-Z0-9|\n\t\s ]+$/';
```
Could we store it to a constant variable?

kumareshinnoraft commented on index.php:
```diff
@@ -0,0 +1,169 @@
+                fclose($fp);
+            } 
+            else {
+                $subjectmarksErr = "Input can only contain alpha-numeric characters and '|'";         
```
Could we move this to a different function and return the messages?

kumareshinnoraft commented on index.php:
```diff
@@ -0,0 +1,169 @@
+            else {
+                $subjectmarksErr = "Input can only contain alpha-numeric characters and '|'";         
+            }
+        }
+
+
+        $phone = $_POST["phone"];
```
Why we are adding too much spaces here?

kumareshinnoraft commented on index.php:
```diff
@@ -0,0 +1,169 @@
+
+        if($phone == ""){
+            $phoneErr = "Phone Number is required";
+        } else {
+            if(!preg_match($indianphone, $phone)) {
```
Could we use `else if` here?

kumareshinnoraft commented on imageForm.php:
```diff
@@ -0,0 +1,60 @@
+        return $data;
+    }
+
+    function __destruct()
+    {
+        // echo "The object of $this->fullname was destroyed";
+    }
```
If we are not using this why we had pushed this code?

kumareshinnoraft commented on imageform.css:
```diff
@@ -0,0 +1,127 @@
+
+tr:hover {
+    background-color: rgba(253, 253, 124, 0.808);
+}
```
We should add an empty line at the end of the file.

kumareshinnoraft commented on imageForm.php:
```diff
@@ -0,0 +1,60 @@
+
+    public $fnameErr, $lnameErr;
+
+    function __construct()
+    {
+        $this->fname = $this->lname = $this->fullname = "";
+        $this->fnameErr = $this->lnameErr = "";
+    }
+
+    function setFirstName(string $fname)
+    {
+        $this->fname = $fname;
+    }
+    function setLastName(string $lname)
+    {
+        $lname = $this->testInput($lname);
+        $this->lname = $lname;
+    }
+
+    function getFirstName(): string
+    {
+        return $this->fname;
+    }
+
+    function getLastName(): string
+    {
+        return $this->lname;
+    }
+
+    function setFullname()
+    {
+        $this->fullname = $this->fname . " " . $this->lname;
+    }
+
+    function getFullName(): string {
+        return $this->fullname;
+    }
```
We should add multiline comments in each function.

kumareshinnoraft commented on index.php:
```diff
@@ -0,0 +1,169 @@
+                echo "<div class='detail'> Phone Number: " . $phone . "</div>";
+            }
+
+            // showing an error message below the image
+            // elseif(isset($_POST["submit"]) && $_POST["sub_marks"] == "") {
+            //     echo "No marks inserted";
+            // }
+
```
We should not push commented code in the PR.

kumareshinnoraft commented on index.php:
```diff
@@ -0,0 +1,169 @@
+            // if form is submitted and no subject|marks were inputted
+            if (isset($_POST["submit"]) && $subjectmarksErr == "") {
+
+                $fp = fopen("marksFile.txt", "r");
+                echo "<table class='tableheading' >";
+                echo "<tr> <td>Subject</td> <td>Marks</td> </tr>";
+                while (!feof($fp)) {
+                    $fileline = fgets($fp);
+                    $fileline = explode("|", $fileline);
+                    echo "<tr>
+            <td>" . $fileline[0] . "</td>
+            <td> " . $fileline[1] . " </td>
+             </tr>";
```
We should avoid using html inside echo.

kumareshinnoraft commented on index.php:
```diff
@@ -0,0 +1,169 @@
+
+            // if the image name is set, we display the image
+            if (isset($img_name) && !empty($img_name)) {
+                echo "<div class='uploadimage'>";
+                echo "<img  src = 'uploads/$img_name'  height=300px>";
+                echo "</div>";
```
We should not use HTML tags inside echo.

kumareshinnoraft reviewed (changes requested):
1. Please add proper comments.
2. Use code reusability.
3. Do not push commented code.
4. Please follow proper standards for commit message.

MD-Aftab-Innoraft committed:
FT2-37-4: Added validation for Indian Phone number field.
- Phone number must strictly consist of 13 characters including '+91'.
- Validated input Phone number against a regex.
- Adding '+91' is mandatory before the 10 digit number.

MD-Aftab-Innoraft committed:
FT2-37-4: Made the Form more visually appealing.
DEscription:
- Displaying the error messages in red.
- Improved spacing between different fields.
